assay/unix.py

```diff
@@ -103,3 +103,47 @@ def events(self):
                 if e.errno != errno.EINTR:
                     raise
 
+
+class KQueue(object):
+    """BSD-based file descriptor polling object that returns objects, not integers."""
+
+    def __init__(self):
+        self.fdmap = {}
+        self.poller = select.kqueue()
+
+    def register(self, obj):
+        fd = obj.fileno()
+        self.fdmap[fd] = obj
+        event = [select.kevent(
+            fd,
+            select.KQ_FILTER_READ,
+            flags=select.KQ_EV_ADD | select.KQ_EV_ENABLE,
+        )]
+        self.poller.control(event, 0)
+
+    def unregister(self, obj):
+        fd = obj.fileno()
+        del self.fdmap[fd]
+        event = [select.kevent(
+            fd,
+            select.KQ_FILTER_READ,
+            select.KQ_EV_DELETE
+        )]
+        self.poller.control(event, 0)
+
+    def events(self):
+        while True:
+            try:
+                for event in self.poller.control(None, 10):
```

Comment:
I picked 10 arbitrarily here; it represents the maximum number of events that can be read. The original commit had this set to 0, which put this into an infinite loop without ever returning anything. Unlike epoll, there doesn't seem to be a way to tell KQueue to "return everything you've got".

Other options I considered:
Set it to 1, and process each event one at a time
Set it to num_cpus so that up to all the waiting workers are processed in one iteration.